# Load Iris

In [177]:
from sklearn.datasets import load_iris
iris=load_iris()

In [178]:
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [179]:
print(iris.data.shape)

(150, 4)


In [180]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

# Split into training and testing sets

In [181]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(iris.data,
iris.target,
random_state=11)

## Create KNN model

In [182]:
from sklearn.neighbors import KNeighborsClassifier 

knn=KNeighborsClassifier()
knn.fit(X=X_train, y=y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [183]:
predicted = knn.predict(X=X_test)

expected=y_test

for i in range (expected.size):
    if(expected[i]!=predicted[i]):
        print((predicted[i],expected[i]))

(1, 2)


In [184]:
from sklearn.metrics import confusion_matrix
confusion= confusion_matrix(y_true=expected, y_pred=predicted)
confusion 

array([[12,  0,  0],
       [ 0, 11,  0],
       [ 0,  1, 14]])

In [185]:
print("Accuracy of KNN: ", f'{knn.score(X_test, y_test):.2%}')

Accuracy of KNN:  97.37%


<h3> Observations on iris dataset </h3>
<h4> The target consists of three classes: setosa, versicolor, virginica denoted by: 0, 1, 2 respectively.
<br> The classes are balanced. </h4>
    
<h3> Observations on the KNN model </h3>
<h4> The KNN model made one mistake only, predicting versicolor (1) when expecting virginica (2). 
<br> And so, its accuracy is high, 97.37%  </h4>

# Hyperparameter Tuning

In [186]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

for k in range(1,20,2):
    kfold=KFold(n_splits=10, random_state=11,shuffle=True)
    knn=KNeighborsClassifier(k)
    scores=cross_val_score(estimator=knn, X=iris.data,y=iris.target,cv=kfold)
    print(f'k={k:>2}: '+ f'mean accuracy={scores.mean():.2%}; ' +f'standard deviation={scores.std():.2%}')

k= 1: mean accuracy=95.33%; standard deviation=5.21%
k= 3: mean accuracy=96.00%; standard deviation=4.42%
k= 5: mean accuracy=95.33%; standard deviation=5.21%
k= 7: mean accuracy=96.67%; standard deviation=5.37%
k= 9: mean accuracy=96.67%; standard deviation=4.47%
k=11: mean accuracy=97.33%; standard deviation=4.42%
k=13: mean accuracy=96.00%; standard deviation=4.42%
k=15: mean accuracy=97.33%; standard deviation=4.42%
k=17: mean accuracy=97.33%; standard deviation=4.42%
k=19: mean accuracy=96.67%; standard deviation=4.47%


<h4> We can see that the mean accuracy of K 11, 15 and 17 are equal, their standard deviation is also equal,
    <br> so they are the optimal k value. </h4>

# Choosing the best estimator

In [187]:
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier

estimators={'KNN':knn, "SVC":SVC(gamma='scale'), 'GaussianNB':GaussianNB(), \
             'Decision Tree': DecisionTreeClassifier(criterion="entropy", max_depth = 4) }

for model,ob in estimators.items():
    kfold=KFold(n_splits=10, random_state=11,shuffle=True)
    scores=cross_val_score(estimator=ob, X=iris.data,y=iris.target,cv=kfold)
    print(f'{model:>20}: '+ f'mean accuracy={scores.mean():.2%}; ' +f'standard deviation={scores.std():.2%}')

                 KNN: mean accuracy=96.67%; standard deviation=4.47%
                 SVC: mean accuracy=94.67%; standard deviation=4.99%
          GaussianNB: mean accuracy=94.67%; standard deviation=6.53%
       Decision Tree: mean accuracy=95.33%; standard deviation=5.21%


<h4> The KNN model has the highest mean accuracy, and so it is the best estimator </h4>